design/build-cache.md

```diff
@@ -0,0 +1,201 @@
+ss# Build cache
+
+## Context
+
+Users require a way to cache resources between jobs within a single pipeline or consecutive executions of a particular pipeline in order to speed up build time.
+
+## Status
+
+- 29 June 2018: Proposal submitted
+
+## Proposal
+
+Provide a mechanism for transparent caching of resources between jobs within a single pipeline, or between consecutive executions of a particular pipeline.
+
+1. Content to be cached can be specified by the user.
+2. Caching is scoped. 
+   1. Job level caching
+   2. Event level caching
+   3. Pipeline level caching (static caching, in a sense)
+3. Well used caching patterns can be offered as predefined options.
+   1. Dependency caching for build systems such as npm, sbt and cargo.
+
+The build cache and cache usage is configurable via Yaml and backed by an internal sd-command and caching service.
+
+## Details
+
+### Yaml specification
+
+```yml
+shared:
+  # Caches defined in the shared section can be pipeline or exection scoped.
+  cache:
+    # A pipeline scoped cache.
+    pipeline:  
+      # The caching targets
+      target: ["build/**", "~/.sbt"]
+    # An event scoped cache
+    event:
+      target: ["target/deployment.jar"]
```

Comment:
While this makes sense as an example for caching files across different builds in the same event, a more common pattern we've been seeing people doing is explicitly publishing the produced artifact to registries (e.g. Artifactory) in one jobs, and pulling it in another for deployment.

Reply:
Yes I thought the `deployment.jar` was the final artifact the pipeline produces and was concerned that the example would encourage people to put these types of resources to cache dir. I think `generated intermediate resources` is probably a better description.

Reply:
## Conclusion

- [ ] Fix `deployment.jar` and related wording to avoid confusion.